epoxy-paging/src/main/java/com/airbnb/epoxy/paging/PagedListModelCache.kt

```diff
@@ -31,130 +32,222 @@ import java.util.concurrent.Executor
         }
-        rebuildCallback()
     }
 
-    override fun onRemoved(position: Int, count: Int) {
-        (0 until count).forEach { _ ->
-          modelCache.removeAt(position)
+    @SuppressLint("RestrictedApi")
+    private val asyncDiffer = object : AsyncPagedListDiffer<T>(
+        updateCallback,
+        AsyncDifferConfig.Builder<T>(
+            itemDiffCallback
+        ).also { builder ->
+            if (diffExecutor != null) {
+                builder.setBackgroundThreadExecutor(diffExecutor)
+            }
+
+            // we have to reply on this private API, otherwise, paged list might be changed when models are being built,
+            // potentially creating concurrent modification problems.
+            builder.setMainThreadExecutor { runnable: Runnable ->
+                modelBuildingHandler.post(runnable)
+            }
+        }.build()
+    ) {
+        init {
+            if (modelBuildingHandler != EpoxyController.defaultModelBuildingHandler) {
+                try {
+                    // looks like AsyncPagedListDiffer in 1.x ignores the config.
+                    // Reflection to the rescue.
+                    val mainThreadExecutorField =
+                        AsyncPagedListDiffer::class.java.getDeclaredField("mMainThreadExecutor")
+                    mainThreadExecutorField.isAccessible = true
+                    mainThreadExecutorField.set(this, Executor {
+                        modelBuildingHandler.post(it)
+                    })
+                } catch (t: Throwable) {
+                    val msg = "Failed to hijack update handler in AsyncPagedListDiffer." +
+                            "You can only build models on the main thread"
+                    Log.e("PagedListModelCache", msg, t)
+                    throw IllegalStateException(msg, t)
+                }
+            }
         }
-        rebuildCallback()
     }
-  }
-
-  private val asyncDiffer = @SuppressLint("RestrictedApi")
-  object : AsyncPagedListDiffer<T>(
-      updateCallback,
-      AsyncDifferConfig.Builder<T>(
-          itemDiffCallback
-      ).also {builder ->
-        if (diffExecutor != null) {
-          builder.setBackgroundThreadExecutor(diffExecutor)
+
+    @Synchronized
+    fun submitList(pagedList: PagedList<T>?) {
+        inSubmitList = true
+        asyncDiffer.submitList(pagedList)
+        inSubmitList = false
+    }
+
+    @Synchronized
+    fun getModels(): List<EpoxyModel<*>> {
+        val currentList = asyncDiffer.currentList ?: emptyList<T>()
+
+        // The first time models are built the EpoxyController does so synchronously, so that
+        // the UI can be ready immediately. To avoid concurrent modification issues with the PagedList
+        // and model cache we can't allow that first build to touch the cache.
+        if (Looper.myLooper() != modelBuildingHandler.looper) {
+            val initialModels = currentList.mapIndexed { position, item ->
```

Comment:
hmm we may end up building more than necessary (or too much) but maybe it is OK since paged list wont load too much.